File: apps/mobile/lib/models/book.dart
import 'package:uuid/uuid.dart';
import 'package:bookshelf/app/constants.dart';

/// Generator UUID global (v4 random).
const _uuid = Uuid();

/// Model data buku.
/// Field-field disesuaikan dengan API backend Go (models/book.go).
class Book {
  /// Primary key — UUID v4 string.
  final String id;

  final String title;
  final String author;
  final String genre;
  final int pageMax;
  int pageCurrent;
  final String? description;
  String? note;
  int? rating;
  ReadingStatus status;
  final String? isbn;
  final String? coverUrl;

  /// Flag sinkronisasi — true jika data sudah ter-sync ke server.
  bool isSynced;

  DateTime updatedAt;

  Book({
    String? id,
    required this.title,
    required this.author,
    required this.genre,
    required this.pageMax,
    this.pageCurrent = 0,
    this.description,
    this.note,
    this.rating,
    this.status = ReadingStatus.belumDibaca,
    this.isbn,
    this.coverUrl,
    this.isSynced = false,
    DateTime? updatedAt,
  }) : id = id ?? _uuid.v4(),
       updatedAt = updatedAt ?? DateTime.now();

  /// Hitung persentase progres baca (0.0 - 1.0).
  double get progressPercent {
    if (pageMax <= 0) return 0.0;
    return (pageCurrent / pageMax).clamp(0.0, 1.0);
  }

  /// Hitung persentase progres sebagai integer (0 - 100).
  int get progressPercentInt => (progressPercent * 100).round();

  /// Konversi dari JSON (response API backend Go).
  /// Field names match backend: models/book.go JSON tags.
  factory Book.fromJson(Map<String, dynamic> json) {
    return Book(
      id: json['id']?.toString(),
      title: json['title'] as String? ?? '',
      author: json['author'] as String? ?? '',
      genre: json['genre'] as String? ?? '',
      pageMax: json['page_max'] as int? ?? 0,
      pageCurrent: json['page_current'] as int? ?? 0,
      description: json['description'] as String?,
      note: json['note'] as String?,
      rating: json['rating'] as int?,
      status: _parseStatus(json['progress'] as String?),
      isbn: json['ISBN'] as String?,
      updatedAt: json['updated_at'] != null
          ? DateTime.tryParse(json['updated_at'].toString())
          : null,
    );
  }

  /// Konversi ke JSON (request API backend Go).
  /// Field names match backend: models/book.go JSON tags.
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'updated_at': updatedAt.toUtc().toIso8601String(),
      'title': title,
      'author': author,
      'genre': genre,
      'page_max': pageMax,
      'page_current': pageCurrent,
      'description': description ?? '',
      'note': note ?? '',
      'rating': rating ?? 0,
      'progress': _statusToApi(status),
      'ISBN': isbn,
    };
  }

  /// Parse status string dari API ke enum ReadingStatus.
  /// Backend values: "belum", "sedang", "selesai"
  static ReadingStatus _parseStatus(String? value) {
    switch (value) {
      case 'sedang':
        return ReadingStatus.sedangDibaca;
      case 'selesai':
        return ReadingStatus.selesai;
      default:
        return ReadingStatus.belumDibaca;
    }
  }

  /// Konversi enum ReadingStatus ke string API.
  static String _statusToApi(ReadingStatus status) {
    switch (status) {
      case ReadingStatus.belumDibaca:
        return 'belum';
      case ReadingStatus.sedangDibaca:
        return 'sedang';
      case ReadingStatus.selesai:
        return 'selesai';
    }
  }
}

File: backend/internal/models/book.go
package models

import (
	"bookshelf/converter"
	"bookshelf/internal/database"
	"time"

	"github.com/google/uuid"
)

type Book struct {
	ID          uuid.UUID `json:"id"`
	UpdatedAt   time.Time `json:"updated_at"`
	Title       string    `json:"title"`
	Author      *string   `json:"author"`
	Genre       *string   `json:"genre"`
	PageMax     *int32    `json:"page_max"`
	PageCurrent *int32    `json:"page_current"`
	Description *string   `json:"description"`
	Note        *string   `json:"note"`
	Rating      *int32    `json:"rating"`
	Progress    *string   `json:"progress"`
	Isbn        *string   `json:"ISBN"`
}

// Database of book have different data type from json book format.
// we need these function to convert to each other
func (book *Book) ToDatabaseFormat(userID uuid.UUID) database.Book {
	return database.Book{
		ID:          book.ID,
		UserID:      userID,
		UpdatedAt:   book.UpdatedAt,
		Title:       book.Title,
		Author:      converter.NewNullString(book.Author),
		Genre:       converter.NewNullString(book.Genre),
		PageMax:     converter.NewNullInt32(book.PageMax),
		PageCurrent: converter.NewNullInt32(book.PageCurrent),
		Description: converter.NewNullString(book.Description),
		Note:        converter.NewNullString(book.Note),
		Rating:      converter.NewNullInt32(book.Rating),
		Progress:    converter.NewNullString(book.Progress),
		Isbn:        converter.NewNullString(book.Isbn),
	}
}

func BookDatabaseToBasicJSONFormat(bookDB database.Book) Book {
	return Book{
		ID:          bookDB.ID,
		UpdatedAt:   bookDB.UpdatedAt,
		Title:       bookDB.Title,
		Author:      &bookDB.Author.String,
		Genre:       &bookDB.Genre.String,
		PageMax:     &bookDB.PageMax.Int32,
		PageCurrent: &bookDB.PageCurrent.Int32,
		Description: &bookDB.Description.String,
		Note:        &bookDB.Note.String,
		Rating:      &bookDB.Rating.Int32,
		Progress:    &bookDB.Progress.String,
		Isbn:        &bookDB.Isbn.String,
	}
}
